Added functional reconnect for mqtt

JohanLeirnes opened the pull request.

JohanLeirnes committed:
move-mqtt-code

Test move mqtt connect before publish.

JohanLeirnes committed:
Test reconnect and loop

JohanLeirnes committed:
added forever

JohanLeirnes committed:
more info in error message

JohanLeirnes committed:
more info on error

JohanLeirnes committed:
Move int ret

JohanLeirnes committed:
Revert to witout _forever

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -202,14 +202,20 @@ class SignalDetector {
-            mosquitto_lib_cleanup();
-            exit(-1);
+          int ret = mosquitto_publish (mosq, NULL, MQTT_TOPIC, strlen(mesg) - 1, mesg, 0, true);
+          mosquitto_loop(mosq, -1, 1);
```
Either remove or read status code..

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -202,14 +202,20 @@ class SignalDetector {
+          if ( ret != MOSQ_ERR_SUCCESS) {
+            mosquitto_reconnect(mosq);
+            ret = mosquitto_publish (mosq, NULL, MQTT_TOPIC, strlen(mesg) - 1, mesg, 0, true);
+            mosquitto_loop(mosq, -1, 1);
```
Either remove or read status code.

fredrike reviewed (changes requested):
I think we should test the `loop_start` call just after mqtt connect and remove the two `loop` calls. I don't think there is any use of them now, but a background thread handling mqtt network communications would be great.

fredrike commented:
This changes do the trick:

```diff
diff --git a/sparsnas_decode.cpp b/sparsnas_decode.cpp
index a447dcc..b2d51ce 100644
--- a/sparsnas_decode.cpp
+++ b/sparsnas_decode.cpp
@@ -202,14 +202,18 @@ public:
       m += sprintf(m, ",\"Sensor\":%6d}\n", SENSOR_ID);
       if (!testing) {
         if (mosq && !bad) {
-          int ret = mosquitto_publish (mosq, NULL, MQTT_TOPIC, strlen(mesg) - 1, mesg, 0, false);
-          if (ret) {
-            fprintf(stderr, "Can't publish to Mosquitto server\n");
-            // Tear down the connecton and exit.
-            mosquitto_disconnect (mosq);
-            mosquitto_destroy (mosq);
-            mosquitto_lib_cleanup();
-            exit(-1);
+          int ret = mosquitto_publish (mosq, NULL, MQTT_TOPIC, strlen(mesg) - 1, mesg, 0, true);
+          if ( ret != MOSQ_ERR_SUCCESS) {
+            mosquitto_reconnect(mosq);
+            ret = mosquitto_publish (mosq, NULL, MQTT_TOPIC, strlen(mesg) - 1, mesg, 0, true);
+            if (ret != MOSQ_ERR_SUCCESS) {
+              fprintf(stderr, "Can't publish to Mosquitto server.\n%s\n", mosquitto_strerror(ret) );
+              // Tear down the connecton and exit.
+              mosquitto_disconnect (mosq);
+              mosquitto_destroy (mosq);
+              mosquitto_lib_cleanup();
+              exit(-1);
+            }
           }
         } else
           bad ? fprintf(stderr, "%s", mesg) : printf("%s", mesg);
@@ -453,6 +457,7 @@ int main(int argc, char **argv)
 
     // Create a new Mosquitto runtime instance with a random client ID,
     mosq = mosquitto_new (NULL, true, NULL);
+    mosquitto_loop_start(mosq);
 
     if (mosq) {
       //Set username and password (will be ignored of MQTT_USERNAME=NULL)
```

JohanLeirnes committed:
Update sparsnas_decode.cpp

Added some crc checks, fixed loop_start and removed loop inside the publish ifstatement

JohanLeirnes commented:
Changed a bit with your input and added some more logic to the crc publishing.

JohanLeirnes committed:
Update sparsnas_decode.cpp

Fixed issues.

JohanLeirnes committed:
Update sparsnas_decode.cpp

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -195,24 +196,30 @@ class SignalDetector {
           error_sum_count += 1;
         }
+      } else {
+      m += sprintf(m, "{\"CRC\": \"ERR\"");
```
It would be better to set the `MQTT_TOPIC` here..

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -195,24 +196,30 @@ class SignalDetector {
+      m += sprintf(m, "{\"CRC\": \"ERR\"");
       }
 
-      m += sprintf(m, (crc == packet_crc) ? ",\"CRC\":\"ok\"" : ",\"CRC\": \"ERR\"");
```
What happens to the json when this is removed and replaced with 

` m += sprintf(m, "{\"CRC\": \"ERR\"");`

JohanLeirnes replied:
The code 
```
m += sprintf(m, "{\"CRC\": \"ERR\"");
```
is in a second if statement. so if crc error the json would look like this {"CRC":"ERR","Sensor":653213} published to the crc topic.
And if not then the json would look like normal published to the normal topic.

JohanLeirnes replied:
<img width="1136" alt="skarmavbild 2018-02-26 kl 11 07 07" src="https://user-images.githubusercontent.com/2168003/36664520-405ceebc-1ae5-11e8-8ff2-ace6e02bad9a.png">

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -440,6 +447,12 @@ int main(int argc, char **argv)
       sprintf(MQTT_TOPIC, "sparsnas/%d", SENSOR_ID);
 
+    memset(MQTT_CRC_TOPIC, '\0', sizeof(MQTT_CRC_TOPIC));
+    if (const char *env_p = std::getenv("MQTT_CRC_TOPIC"))
```
This is not needed, I think  it is better to let `MQTT_CRC_TOPIC` be:

`sprintf(MQTT_CRC_TOPIC, "%s/crc", MQTT_TOPIC, SENSOR_ID);`

JohanLeirnes replied:
This is true.

JohanLeirnes committed:
Update sparsnas_decode.cpp

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -195,24 +196,30 @@ class SignalDetector {
+          mosquitto_publish (mosq, NULL, MQTT_CRC_TOPIC, strlen(mesg) - 1, mesg, 0, true);
         } else
-          bad ? fprintf(stderr, "%s", mesg) : printf("%s", mesg);
+        bad ? fprintf(stderr, "%s", mesg) : printf("%s", mesg);
```
Please add the spaces again (the line 222 is part of the else on lin 221).

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -195,24 +196,30 @@ class SignalDetector {
 
-      m += sprintf(m, (crc == packet_crc) ? ",\"CRC\":\"ok\"" : ",\"CRC\": \"ERR\"");
-
       m += sprintf(m, ",\"Sensor\":%6d}\n", SENSOR_ID);
```
insert this:

```c
char* topic = (crc == packet_crc) ? MQTT_TOPIC : MQTT_CRC_TOPIC;
```

And change `MQTT_TOPIC` below to `topic`

JohanLeirnes committed:
Update sparsnas_decode.cpp

JohanLeirnes committed:
Update sparsnas_decode.cpp

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -195,21 +195,26 @@ class SignalDetector {
-            mosquitto_destroy (mosq);
-            mosquitto_lib_cleanup();
-            exit(-1);
+        if (mosq && !bad && crc == packet_crc) {
```
` && crc == packet_crc`is redundant.

JohanLeirnes committed:
Update sparsnas_decode.cpp

JohanLeirnes committed:
Update sparsnas_decode.cpp

debugging for effect.
Also changed watt to double from float

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -178,38 +178,43 @@ class SignalDetector {
         bad = false;
         int seq = (dec[9] << 8 | dec[10]);
-        int effect = (dec[11] << 8 | dec[12]);
+        unsigned int effect = (dec[11] << 8 | dec[12]);
```
Please revert this back to just an `int` (as it didn't affect the "watt").

fredrike commented on sparsnas_decode.cpp:
```diff
@@ -178,38 +178,43 @@ class SignalDetector {
 //      Note that data_[4] cycles between 0-3 when you first put in the batterys in t$
         if(data4 == 1){
-          watt = (float)((3600000 / PULSES_PER_KWH) * 1024) / (effect);
+          watt = (double)((3600000 / PULSES_PER_KWH) * 1024) / (effect);
```
Please revert back to `float` (as it didn't affect the "watt").
